### PR TITLE
Python debug scripts

### PR DESCRIPTION
**Pull request by mbolivar**
TL;DR
=====

This series finishes rewriting the flash and debug scripts in Python, which should help the CMake switch a bit. Since there were a few issues last time, there are more details below.  **JLink on frdm_k64f has an issue here and on master -- just me?**

It also fixes two minor issues with the Python script (in "scripts: zephyr_flash_debug: add missing nrfjprog line" and "scripts: zephyr_flash_debug: fix ARC OPENOCD_EXTRA_INIT").

Casting a wide net of reviewers among build, arch, and SoC people just in case.

Details
=======

Results for testing Python equivalents to shell scripts follow. Linux-specific shell idioms were converted to portable alternatives for macOS and Windows targets, but this series has only been tested end-to-end on Linux.

Flashing
--------------

Not tested on HW (I don't have boards):

- bossa-flash.sh
- esp32.sh
- nios2.sh

Tested that flashing hello world gives console output:

- dfuutil.sh: 96b_carbon,  arduino_101
- pyocd.sh: frdm_k64f, 96b_nitrogen, nrf52_blenano2
- nrf_flash.sh: nrf52840_pca10056
- openocd.sh: disco_l475_iot1

Tested to run without error only (I don't know how to check the console to make sure the right chip was flashed):

- dfuutil.sh: arduino_101_sss, arduino_101_ble

Debugging
---------------

Not tested on HW (I don't have boards, or it's QEMU):

- arc_debugger.sh
- nios2.sh
- xt-gdb.sh
- qemu.sh

Tested that 'debug' connects and allows resetting and running code:

- openocd.sh: disco_l475_iot1
- pyocd.sh: frdm_k64f, 96b_nitrogen, nrf52_blenano2

**Issue**

- jlink.sh: on frdm_k64f, I tried `make -j8 debug BOARD=frdm_k64f OPENSDA_FW=jlink`. This failed, but this doesn't seem to be a regression; I get the same error on master (https://hastebin.com/ijoxuhaxom.sql). Is it just me?

Other tests
-----------

I hacked up the Python and shell scripts on another branch to log their output to files instead (with some filtering to make the shell versus Python diffs less noisy):

https://github.com/mbolivar/zephyr/tree/python-debug-scripts-testing

On that branch, I ran both flash and debug for every Zephyr board samples/hello_world built on (esp32 and xt-sim had issues using the 0.9.2-rc5 SDK) and verified the logs were equivalent.

Future
------

- make the Python script smarter (the first goal is adding DT-awareness for correct flash offsets when relevant)
- make this less messy (use arguments instead of environment variables), maybe clean up the Makefile.board files accordingly (might just punt this in favor of CMake)

### CODE REVIEW

File: scripts/support/zephyr_flash_debug.py [hunk 1]
```diff
@@ -177,71 +312,77 @@ def create_from_env(debug):
...
         self.bin_name = bin_name
         self.bossac = bossac
 
-    def replaces_shell_script(shell_script):
-        return shell_script == 'bossa-flash.sh'
+    def replaces_shell_script(shell_script, command):
+        return command == 'flash' and shell_script == 'bossa-flash.sh'
 
-    def create_from_env(debug):
+    def create_from_env(command, debug):
```
Comment (galak) [inline scripts/support/zephyr_flash_debug.py]: Since you are cleaning all things up, can you look at adding in when I did with PR #4502

File: Makefile [hunk 2]
```diff
@@ -1390,9 +1390,13 @@ flash: FORCE
 endif
 
 ifneq ($(DEBUG_SCRIPT),)
+ifeq ($(USE_ZEPHYR_FLASH_DEBUG_SHELL),)
```
Comment (galak) [inline Makefile]: is there a reason not to just use your script always?

File: scripts/support/zephyr_flash_debug.py [hunk 3]
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
...
+
+    def _parser_linux_darwin(self, cmd):
+        out = subprocess.check_output(cmd).split(b'\n')
+        used_bytes = [x.split()[3].rsplit(b':', 1)[1] for x in out
+                      if x.startswith(b'tcp')]
+        return {int(b) for b in used_bytes}
+
+
+class ZephyrBinaryRunner(abc.ABC):
+    '''Abstract superclass for binary runners (flashers, debuggers).'''
```
Comment (galak) [inline scripts/support/zephyr_flash_debug.py]: Can you add more docs about the theory/practice of how this class would be utilized.

**Review by galak — CHANGES_REQUESTED** [review]
What about refactor this a little so that each of the runners are in their own files, and make it a bit more library so other script front ends might utilize the same bits.

I'm thinking about how we might write or modify sanitycheck to allow running tests built on a local board.

**Comment by mbolivar** [comment]
> What about refactor this a little so that each of the runners are in their own files, and make it a bit more library so other script front ends might utilize the same bits.

Sure, that's fine. Can I do that on top of this series since it already starts out with them all in one file?

File: Makefile (hunk 2 above)
Comment (mbolivar) [reply Makefile]: With this series, the script is being used by default now for everything (flash, debug, debugserver).

I'm leaving this in as an escape hatch in case there are problems after the merge, the same way as was done when I re-did the flashing scripts in https://github.com/zephyrproject-rtos/zephyr/pull/4304.

My plan is to remove the shell scripts once these get merged and a few days of testing, but I an remove them at the end of this series as well if that's better.

File: scripts/support/zephyr_flash_debug.py (hunk 1 above)
Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: Sounds good! That class was a pretty primitive one.

File: scripts/support/zephyr_flash_debug.py (hunk 3 above)
Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: Sounds good.

File: scripts/support/zephyr_flash_debug.py [hunk 4]
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
...
+    '''Spawn a child command, ensuring it ignores SIGINT.
+
+    The returned subprocess.Popen object must be manually terminated.'''
+    cflags = 0
+    preexec = None
+    system = platform.system()
+
+    if system == 'Windows':
+        cflags |= subprocess.CREATE_NEW_PROCESS_GROUP
+    elif system in {'Linux', 'Darwin'}:
```
Comment (lpereira) [inline scripts/support/zephyr_flash_debug.py]: What's the value of `os.setsid` on Windows? Might be the case that, if it's available, use it.

File: scripts/support/zephyr_flash_debug.py [hunk 5]
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
...
-        print(' '.join(cmd))
+        print(quote_sh_list(cmd))
     subprocess.check_call(cmd)
 
 
 def check_output(cmd, debug):
     if debug:
-        print(' '.join(cmd))
+        print(quote_sh_list(cmd))
     return subprocess.check_output(cmd)
```
Comment (lpereira) [inline scripts/support/zephyr_flash_debug.py]: Are we doing the shell escaping ourselves because of Windows? Does it have anything similar to `execv()` that can be used instead?

File: scripts/support/zephyr_flash_debug.py [hunk 6]
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
...
+        }
+        handler = handlers[platform.system()]
+        return handler()
+
+    def _used_now_windows(self):
+        cmd = ['netstat', '-a', '-n', '-p', 'tcp']
+        return self._parser_windows(cmd)
+
+    def _used_now_linux(self):
+        cmd = ['netstat', '-a', '-n', '-t']
```
Comment (lpereira) [inline scripts/support/zephyr_flash_debug.py]: `netstat` has been deprecated in Linux; my system at home doesn't even have it installed.  The `ss` utility from `iproute2` is the replacement. It might be the case both need to be tried.

File: Makefile (hunk 2 above)
Comment (andrewboie) [reply Makefile]: it's straightforward enough to revert a PR if it causes problems.
problem with checks like this, in my experience they don't always get removed and become cruft.

**Comment by andrewboie** [comment]
I concur with @galak I would like to see the arch or platform specific script logic separated out from the main script.

File: scripts/support/zephyr_flash_debug.py (hunk 3 above)
Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: Sure; I've tried to improve this

**Comment by mbolivar** [comment]
v2 is posted now addressing comments from @galak. Changes from v1:

- **Delete all shell scripts** in a patch at the end
- Add a couple of debug= arguments to constructors where they were missing (not sure how this got past my flake8 test last time)
- Update the bossac runner to take a serial port from  the environment
- Add a patch which refactors the unified script into a shim on top of a library with individual files per runner
- Add more documentation and theory of ops to ZephyrBinaryRunner

File: scripts/support/zephyr_flash_debug.py (hunk 5 above)
Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: This uses the standard Python module shlex behind the scenes. In addition to making this portable for Windows, this also means that the output can be copy and pasted and the shell tokenization wil remain what the command in list form intended.

File: scripts/support/zephyr_flash_debug.py (hunk 6 above)
Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: Hrm, OK. Use of netstat matches the existing shell script, which I'm trying to replicate fairly closely in this series (except of course where it needs to be changed for non-Linux targets).

Would you be OK with me fixing this to use ss in a follow-up? I can do it now if that's better, but it feels like the wrong series to do it in.

**Comment by mbolivar** [comment]
@andrewboie, done.

@lpereira, I have responded to your comments; please let me know if you'd like further changes along those lines.

File: Makefile (hunk 2 above)
Comment (mbolivar) [reply Makefile]: I went ahead and deleted all the shell scripts in v2.

File: scripts/support/zephyr_flash_debug.py (hunk 5 above)
Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: Looking at this again, I think you might be asking about the subprocess.check_output line, not quote_sh_list(). If that's right, I'm not sure what you mean by "doing the shell escaping ourselves". The calling conventions for subprocess routines recommend using a sequence for reasons explained here:

https://docs.python.org/3/library/subprocess.html#frequently-used-arguments

Given that cmd is a list, this looks very similar to execv(), so I think it's doing what you're looking for. Please LMK if not.

**Review by tyrantlizard77 — APPROVED** [review]
approach taken here looks reasonable to me

File: scripts/support/zephyr_flash_debug.py (hunk 6 above)
Comment (lpereira) [reply scripts/support/zephyr_flash_debug.py]: It can be done in a follow up patch, no problems.

Comment (mbolivar) [reply scripts/support/zephyr_flash_debug.py]: Thanks, I'll queue this up in the next series:

https://github.com/mbolivar/zephyr/commit/57e096f165d6c59affb42d719d69bb71fb51dc0c

**Comment by mbolivar** [comment]
v3: rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/4502

**Review by galak — CHANGES_REQUESTED** [review]
Can you move the refactor script/lib patch before adding the new classes.

**Comment by mbolivar** [comment]
> Can you move the refactor script/lib patch before adding the new classes.

Done. Diff from the previous version is just deleting an extraneous import caught by flake8: https://hastebin.com/nehularafa.sql

**Comment by mbolivar** [comment]
One more try. Botched the the refactor __init__.py by importing classes that didn't get added until later. Same end result, but intermediate commits are OK now.

**Review by nashif — APPROVED** [review]
